mirrord-layer/src/pod_api.rs

```diff
@@ -412,8 +392,345 @@ async fn create_job_pod_agent(
+                })
+            }
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is neither a pod or a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl RuntimeDataProvider for PodData {
+    async fn runtime_data(&self, client: &Client, namespace: &str) -> Result<RuntimeData> {
+        let pod_api: Api<Pod> = Api::namespaced(client.clone(), namespace);
+        let pod = pod_api.get(&self.pod_name).await?;
+        let (
+            ContainerRuntime {
+                container_runtime,
+                socket_path,
+                container_id,
+            },
+            container_name,
+        ) = || -> Option<(String, String)> {
+            let container_statuses = &pod.clone().status?.container_statuses?;
+            let container_name = if let Some(container_name) = self.container_name.clone() {
+                container_name
+            } else {
+                pod.clone()
+                    .spec?
+                    .containers
+                    .first()
+                    .map(|container| container.name.clone())?
+            };
+            let container_info = container_statuses
+                .iter()
+                .find(|&status| status.name == container_name)?
+                .container_id
+                .clone()?;
+            Some((container_info, container_name))
+        }()
+        .ok_or_else(|| {
+            LayerError::ContainerNotFound(format!(
+                "Failed to fetch container for pod {:#?}",
+                self.clone(),
+            ))
+        })
+        .and_then(|(runtime_and_id, container_name)| {
+            Ok((runtime_and_id.parse::<ContainerRuntime>()?, container_name))
+        })?;
+
+        let node_name = pod
+            .spec
+            .and_then(|spec| spec.node_name)
+            .ok_or_else(|| LayerError::NodeNotFound(format!("{:?}", self.clone())))?;
+
+        Ok(RuntimeData {
+            container_id,
+            container_runtime,
+            node_name,
+            socket_path,
+            pod_name: self.pod_name.clone(),
+            container_name,
+        })
+    }
+}
+
+impl FromStr for Target {
+    type Err = LayerError;
+
+    fn from_str(target: &str) -> Result<Self> {
+        target
+            .parse::<DeploymentData>()
+            .map(Target::Deployment)
+            .or_else(|_| target.parse::<PodData>().map(Target::Pod))
+    }
+}
+
+pub(crate) struct ContainerRuntime {
+    container_runtime: String,
+    socket_path: String,
+    container_id: String,
+}
+
+impl FromStr for ContainerRuntime {
+    type Err = LayerError;
+    fn from_str(input: &str) -> Result<Self> {
+        let container_runtime_info = input.split("://").collect::<Vec<&str>>();
+        let (container_runtime, socket_path) = match container_runtime_info.first() {
+            Some(&"docker") => ("docker", "/var/run/docker.sock"),
```

Comment:
Why does this need to populate `container_runtime` with the same value that's already in `container_runtime_info.first()`?